### PR TITLE
Fixes #39313 - Fix RHSM owner proxy authorization and owner label handling

### PR DESCRIPTION
**Pull request by nofaralfasi**
Remove controller-instance authorization checks for owner proxy routes and normalize organization labels to IDs for proxy actions so /rhsm/owners/:organization_id/* works without requiring an owner query param. Add controller coverage for both authorization paths and organization label conversion.

#### What are the changes introduced in this pull request?
- Updated RHSM proxy behavior to normalize `organization_id` labels for proxy get action in addition to existing actions, so `/rhsm/owners/:organization_id/system_purpose` works without requiring `?owner=....`
- Added a guard in `convert_organization_label_to_id` to leave numeric organization IDs unchanged.
- Expanded controller test coverage:
- label-to-id conversion tests (label conversion + numeric pass-through),
- owner-pools authorization tests for users with/without `view_organizations`,
- end-to-end-style controller test that verifies `system_purpose` works without owner query param.

#### Considerations taken when implementing this change?
- Kept behavior backward compatible by not converting already numeric `organization_id` values.
- Scoped the conversion change to relevant proxy actions only, minimizing side effects.
- Preserved and validated route-level authorization semantics for non-consumer users.
- Added regression tests for both logic-level and request-level paths to prevent future breakage of the same user flow.

#### What are the testing steps for this pull request?
- Create a non-admin user with `view_organizations` permission and org (`Default Organization`) access.
- Call the endpoint without owner query param:
- curl -sku "view_organizations_user:password" "https://<host>/rhsm/owners/Default_Organization/system_purpose"
- Verify response is successful and contains system purpose JSON (no Organization with id ... not found).
- (Optional negative check) Use a user without view_organizations; verify access is denied for owner routes.
- Run controller tests:
`test/controllers/api/rhsm/candlepin_proxies_controller_test.rb`

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved proxy API handling of `organization_id` by converting organization labels to numeric IDs when needed, including the `get` action, while preserving already-numeric values.
  * Corrected authorization behavior for owner-related proxy routes when the consumer parameter is omitted.
  * Fixed system-purpose proxy requests to use the owner label in the URL path (not an `owner` query parameter), returning the expected owner key.
* **Tests**
  * Added controller-spec coverage for `organization_id` conversion, the updated owner-route authorization rules, and system-purpose request formatting.

### CODE REVIEW

File: app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb [hunk 1]
```diff
@@ -382,6 +382,8 @@ def find_organization(key = :organization_id)
     end
 
     def convert_organization_label_to_id
+      return if params[:organization_id].to_s.match?(/\A\d+\z/)
```
Comment (jeremylenz) [inline app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb]: So this is skipping "find org by label" if you send a numeric ID and not a label, right?

1. would be great to have a comment explaining this. I don't like deciphering regex 😄 
2. When is this problem actually encountered? Candlepin only accepts labels, not numeric IDs. (The numeric ID would be the Foreman database ID.) As far as I know, subscription-manager will never send a request with a numeric ID.

File: app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb [hunk 2]
```diff
@@ -31,7 +31,7 @@ class Api::Rhsm::CandlepinProxiesController < Api::V2::ApiController
     before_action :check_media_type, :except => :async_hypervisors_update
 
     prepend_before_action :convert_owner_to_organization_id, :except => [:hypervisors_update, :async_hypervisors_update], :if => lambda { params.key?(:owner) }
-    prepend_before_action :convert_organization_label_to_id, :only => [:rhsm_index, :consumer_activate, :consumer_create], :if => lambda { params.key?(:organization_id) }
+    prepend_before_action :convert_organization_label_to_id, :only => [:rhsm_index, :consumer_activate, :consumer_create, :get, :post, :put, :delete], :if => lambda { params.key?(:organization_id) }
```
Comment (jeremylenz) [inline app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb]: Why did we need to add these?

**Review by jeremylenz — COMMENTED** [review]
hi! Just a couple questions for my own understanding.

File: app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb (hunk 2 above)
Comment (nofaralfasi) [reply app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb]: You are right, we don’t need all 4 for this issue.
I checked routes, and the :organization_id owner proxy endpoints are all GET:
`/owners/:organization_id/pools -> candlepin_proxies#get`
`/owners/:organization_id/servicelevels -> candlepin_proxies#get`
`/owners/:organization_id/system_purpose -> candlepin_proxies#get`
So I narrowed the callback scope to just :get (plus existing non-proxy actions), which keeps the fix minimal and targeted.

File: app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb (hunk 1 above)
Comment (nofaralfasi) [reply app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb]: 1. Re: regex readability
Agreed. I’ll add a short inline comment explaining that this skips label lookup when organization_id is already a numeric Foreman org ID.

2. This can happen in controller callback interaction: `convert_owner_to_organization_id` can set `params[:organization_id]` to the Foreman org ID. Since this PR adds label normalization to owner proxy request handling, we need to avoid re-resolving an already-numeric value as a label (which would fail).

Comment (jeremylenz) [reply app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb]: Edge case: won't this break things if you have a numeric organization label? I don't know of any rules against that.

**Comment by jeremylenz** [comment]
test failures seem related now

File: app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb [hunk 3]
```diff
@@ -383,6 +383,19 @@ def find_organization(key = :organization_id)
...
+      organization_id = params[:organization_id].to_s
+      raise unless organization_id.match?(/\A\d+\z/)
+
+      organization = Organization.find_by(:id => organization_id)
+      raise unless organization
+
+      if User.current && !User.consumer? && !User.current.allowed_organizations.include?(organization)
+        message = _("User '%{user}' does not belong to Organization '%{organization}'.")
+        raise HttpErrors::NotFound, message % {:user => current_user.login, :organization => organization_id}
+      end
```
Comment (jeremylenz) [inline app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb:401]:
```suggestion
      end
    params[:organization_id] = organization.id
```
To make sure params[:organization_id] stays a string, we should have a final return value here.

File: app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb [hunk 4]
```diff
@@ -383,6 +383,19 @@ def find_organization(key = :organization_id)
...
     def convert_organization_label_to_id
       params[:organization_id] = find_organization.id
+    rescue HttpErrors::NotFound
+      # If label lookup fails and organization_id is numeric, treat it as a
+      # Foreman org ID set by earlier callbacks.
+      organization_id = params[:organization_id].to_s
+      raise unless organization_id.match?(/\A\d+\z/)
+
+      organization = Organization.find_by(:id => organization_id)
+      raise unless organization
```
Comment (jeremylenz) [inline app/controllers/katello/api/rhsm/candlepin_proxies_controller.rb]: Could potentially provide a better error message here since `Couldn't find Organization '5'` (the original message, referencing a label) is slightly misleading when the numeric ID doesn't exist either.

**Review by jeremylenz — COMMENTED** [review]
just a couple more nits

**Review by jeremylenz — APPROVED** [review]
Thanks @nofaralfasi!

ACK 👍
